ruby/ql/src/queries/security/cwe-020/IncompleteHostnameRegExp.ql

```diff
@@ -0,0 +1,55 @@
+/**
+ * @name Incomplete regular expression for hostnames
+ * @description Matching a URL or hostname against a regular expression that contains an unescaped dot as part of the hostname might match more hostnames than expected.
+ * @kind problem
+ * @problem.severity warning
+ * @security-severity 7.8
+ * @precision high
+ * @id rb/incomplete-hostname-regexp
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-20
+ */
+
+private import ruby
+private import codeql.ruby.ApiGraphs
+
+private string commonTopLevelDomainRegex() { result = "com|org|edu|gov|uk|net|io" }
+
+/**
+ * Holds if `pattern` is a regular expression pattern for URLs with a host matched by `hostPart`,
+ * and `pattern` contains a subtle mistake that allows it to match unexpected hosts.
+ */
+bindingset[pattern]
+predicate isIncompleteHostNameRegExpPattern(string pattern, string hostPart) {
+  hostPart =
+    pattern
+        .regexpCapture("(?i).*" +
+            // an unescaped single `.`
+            "(?<!\\\\)[.]" +
+            // immediately followed by a sequence of subdomains, perhaps with some regex characters mixed in, followed by a known TLD
+            "([():|?a-z0-9-]+(\\\\)?[.](" + commonTopLevelDomainRegex() + "))" + ".*", 1)
+}
+
+predicate regularExpression(Expr n, string pattern) {
+  (
+    n instanceof RegExpLiteral
+    or
+    exists(API::Node regexp, DataFlow::CallNode callNode |
+      regexp = API::getTopLevelMember("Regexp") and
+      (callNode = regexp.getAnInstantiation() or callNode = regexp.getAMethodCall("compile")) and
+      n = callNode.getArgument(0).asExpr().getExpr()
+    )
```

Comment:
I notice this disjunct is almost identical to [this charpred](https://github.com/github/codeql/blob/1a91a79b5b503f51ddc6ed2eb53b70b8189e897f/ruby/ql/lib/codeql/ruby/security/performance/RegExpInjectionCustomizations.qll#L56-L60). I can't quite decide whether or not it's complicated enough to warrant having a shared predicate between the two.

Reply:
Yes, I wondered the same.